internal/controller/gateway.go

```diff
@@ -31,46 +37,201 @@ import (
+func (r *gatewayReconciler) isController(gc *gatewayapi_v1alpha1.GatewayClass) bool {
+	return gc.Spec.Controller == r.classController
+}
+
 func (r *gatewayReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.log.WithField("namespace", request.Namespace).WithField("name", request.Name).Info("reconciling gateway")
 
-	// Fetch the Gateway from the cache.
-	gateway := &gatewayapi_v1alpha1.Gateway{}
-	err := r.client.Get(ctx, request.NamespacedName, gateway)
-	if errors.IsNotFound(err) {
-		r.eventHandler.OnDelete(&gatewayapi_v1alpha1.Gateway{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      request.Name,
-				Namespace: request.Namespace,
-			},
-		})
+	// Fetch the Gateway.
+	gw := &gatewayapi_v1alpha1.Gateway{}
+	if err := r.client.Get(ctx, request.NamespacedName, gw); err != nil {
+		if errors.IsNotFound(err) {
+			r.log.WithField("name", request.Name).WithField("namespace", request.Namespace).Info("failed to find gateway")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object, so requeue the request.
+		return reconcile.Result{}, fmt.Errorf("failed to get gateway %s/%s: %w", request.Namespace, request.Name, err)
+	}
+
+	// Check if object is deleted.
+	if !gw.ObjectMeta.DeletionTimestamp.IsZero() {
+		r.eventHandler.OnDelete(gw)
+		if err := r.ensureGatewayDeleted(gw); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to ensure deletion of gateway %s/%s: %w",
+				gw.Namespace, gw.Name, err)
+		}
 		return reconcile.Result{}, nil
 	}
 
-	// Pass the new changed object off to the eventHandler.
-	r.eventHandler.OnAdd(gateway)
+	// The gateway is safe to process, so check if it's valid.
+	errs := validation.ValidateGateway(ctx, r.client, gw)
+	if errs != nil {
+		r.log.WithField("name", request.Name).WithField("namespace", request.Namespace).
+			Error("invalid gateway: ", int_errors.ParseFieldErrors(errs))
+	} else {
+		// TODO: Finalize the gateway when managed infrastructure is added.
+
+		// If needed, finalize the referenced gatewayclass.
+		if !r.isClassFinalized() {
+			if err := r.ensureClassFinalizer(); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to finalize gatewayclass %s: %w",
+					r.referencedClass.Name, err)
+			}
+			r.log.WithField("name", r.referencedClass.Name).Info("finalized gatewayclass")
+			// The gatewayclass has been mutated, so get the latest.
+			key := types.NamespacedName{
+				Namespace: r.referencedClass.Namespace,
+				Name:      r.referencedClass.Name,
+			}
+			if err := r.client.Get(ctx, key, r.referencedClass); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to get gatewayclass %s: %w",
+					r.referencedClass.Name, err)
+			}
+		}
+		// Pass the gateway off to the eventHandler.
+		r.eventHandler.OnAdd(gw)
+
+		// TODO: Ensure the gateway by creating manage infrastructure, i.e. the Envoy service.
+	}
+
+	if err := status.SyncGateway(ctx, r.client, gw, errs); err != nil {
```

Comment:
Is this status Sync pattern common for controller-runtime controllers? In any case, we need to ensure that the `Scheduled` Condition is populated correctly (it signals that Contour has accepted the Gateway for processing and, eventually, started the process of creating infrastructure). We should add `Ready` with `status: false` when we add `scheduled`.

Then, `Ready` can be set to `status: true` when the infrastructure is provisioned.

I think we may need to change `SyncGateway()` to update `Scheduled` first maybe?

Reply:
> Is this status Sync pattern common for controller-runtime controllers?

It's a pattern I have seen with other controllers, essentially reconcile the object and calculate status before updating the object on the API server.

> I think we may need to change SyncGateway() to update Scheduled first maybe?

Ack. I was still mentally operating in the operator world where the infra is provisioned. I have updated `computeGatewayReadyCondition()` to `computeGatewayScheduledCondition()` to manage the "Scheduled" condition type. When Envoy infra management is added, we can re-introduce `computeGatewayReadyCondition()` to calculate the "Ready" condition type.